8309613: [Windows] hs_err files sometimes miss information about the code containing the error

TheRealMDoerr opened the pull request:
We have seen hs_err files for errors triggered by C2 compiled methods which miss the most relevant information: the C2 method (see JBS issue for more details). I have found a possibility to add it. Please take a look and provide feedback.

Testing:
```diff
diff --git a/src/hotspot/share/opto/parse1.cpp b/src/hotspot/share/opto/parse1.cpp
index f179d3ba88d..c35a1ac595e 100644
--- a/src/hotspot/share/opto/parse1.cpp
+++ b/src/hotspot/share/opto/parse1.cpp
@@ -1210,6 +1210,12 @@ void Parse::do_method_entry() {
     make_dtrace_method_entry(method());
   }
 
+  if (UseNewCode) {
+    Node* halt = _gvn.transform(new HaltNode(control(), frameptr(), "Requested Halt!"));
+    C->root()->add_req(halt);
+    set_control(halt);
+  }
+
 #ifdef ASSERT
   // Narrow receiver type when it is too broad for the method being parsed.
   if (!method()->is_static()) {
```

"java -XX:+UseNewCode -version" shows the following output (when no hsdis lib is provided):
```
---------------  T H R E A D  ---------------

Current thread (0x0000024daebb2b30):  JavaThread "main"             [_thread_in_Java, id=30876, stack(0x000000cdacc00000,0x000000cdacd00000) (1024K)]

Stack: [0x000000cdacc00000,0x000000cdacd00000]
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [jvm.dll+0x6ca5b9]  os::win32::platform_print_native_stack+0xd9  (os_windows_x86.cpp:236)
V  [jvm.dll+0x8a3af1]  VMError::report+0xd61  (vmError.cpp:991)
V  [jvm.dll+0x8a5d6e]  VMError::report_and_die+0x5fe  (vmError.cpp:1797)
V  [jvm.dll+0x283061]  report_fatal+0x71  (debug.cpp:212)
V  [jvm.dll+0x621c3e]  MacroAssembler::debug64+0x8e  (macroAssembler_x86.cpp:829)
C  0x0000024dc1553cf4

The last pc belongs to nmethod (printed below).

Compiled method (c2)      92   16       4       java.lang.Object::<init> (1 bytes)
 total in heap  [0x0000024dc1553b10,0x0000024dc1553d50] = 576
 relocation     [0x0000024dc1553c70,0x0000024dc1553c88] = 24
 main code      [0x0000024dc1553ca0,0x0000024dc1553d00] = 96
 stub code      [0x0000024dc1553d00,0x0000024dc1553d18] = 24
 metadata       [0x0000024dc1553d18,0x0000024dc1553d20] = 8
 scopes data    [0x0000024dc1553d20,0x0000024dc1553d28] = 8
 scopes pcs     [0x0000024dc1553d28,0x0000024dc1553d48] = 32
 dependencies   [0x0000024dc1553d48,0x0000024dc1553d50] = 8

[Constant Pool (empty)]

[MachCode]
[Entry Point]
  # {method} {0x0000000800478d78} '<init>' '()V' in 'java/lang/Object'
  #           [sp+0x20]  (sp of caller)
  0x0000024dc1553ca0: 448b 5208 | 49bb 0000 | 0000 0800 | 0000 4d03 | d349 3bc2 

  0x0000024dc1553cb4: ;   {runtime_call ic_miss_stub}
  0x0000024dc1553cb4: 0f85 c6a9 | 8fff 6690 | 0f1f 4000 
[Verified Entry Point]
  0x0000024dc1553cc0: 4881 ec18 | 0000 0048 | 896c 2410 | 4181 7f20 | 0100 0000 | 0f85 1b00 

  0x0000024dc1553cd8: ;   {external_word}
  0x0000024dc1553cd8: 0000 48b9 | 98e7 0702 | fd7f 0000 | 4883 e4f0 

  0x0000024dc1553ce8: ;   {runtime_call MacroAssembler::debug64}
  0x0000024dc1553ce8: 48b8 b01b | c201 fd7f | 0000 ffd0 

  0x0000024dc1553cf4: ;   {runtime_call StubRoutines (final stubs)}
  0x0000024dc1553cf4: f4e8 0622 | 8eff e9db | ffff fff4 
[Exception Handler]
  0x0000024dc1553d00: ;   {no_reloc}
  0x0000024dc1553d00: e97b c99d | ffe8 0000 | 0000 4883 

  0x0000024dc1553d0c: ;   {runtime_call DeoptimizationBlob}
  0x0000024dc1553d0c: 2c24 05e9 | 0c40 90ff | f4f4 f4f4 
[/MachCode]


---------------  P R O C E S S  ---------------
```

Without the patch, we don't see which compiled method called this at all. An example from a real C2 bug is here: https://bugs.openjdk.org/secure/attachment/103766/hs_err_pid7652.log

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309613](https://bugs.openjdk.org/browse/JDK-8309613): [Windows] hs_err files sometimes miss information about the code containing the error (**Bug** - P3)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [56a3f4c6](https://git.openjdk.org/jdk/pull/14358/files/56a3f4c69fd7272b9bd9a79873b282c27bbf897b)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [56a3f4c6](https://git.openjdk.org/jdk/pull/14358/files/56a3f4c69fd7272b9bd9a79873b282c27bbf897b)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14358/head:pull/14358` \
`$ git checkout pull/14358`

Update a local copy of the PR: \
`$ git checkout pull/14358` \
`$ git pull https://git.openjdk.org/jdk.git pull/14358/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14358`

View PR using the GUI difftool: \
`$ git pr show -t 14358`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14358.diff">https://git.openjdk.org/jdk/pull/14358.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14358#issuecomment-1580967494)

dholmes-ora commented on src/hotspot/share/utilities/vmError.cpp:
```diff
@@ -675,6 +675,12 @@ void VMError::report(outputStream* st, bool _verbose) {
   static char buf[O_BUFLEN];
 
+  // keep track of which code has already been printed
+  const int printed_capacity = max_error_log_print_code;
+  address printed[printed_capacity];
```
Does this buffer get reused/overwritten by the "printing code blobs" logic?

TheRealMDoerr replied:
The purpose of the buffer is to keep track of what has already been printed. We don't want to print the same code again if we encounter the address again. This can be used across several error reporting steps.

dholmes-ora replied:
Okay, I just wanted to make sure the different usages don't interfere with each other.

dholmes-ora commented on src/hotspot/share/utilities/vmError.cpp:
```diff
@@ -963,9 +969,15 @@ void VMError::report(outputStream* st, bool _verbose) {
+    if (os::platform_print_native_stack(st, _context, buf, sizeof(buf), lastpc)) {
       // We have printed the native stack in platform-specific code
       // Windows/x64 needs special handling.
+      // Stack walking may got stuck. Try to print the calling code.
```
Nit: s/got/get/

TheRealMDoerr replied:
Done.

dholmes-ora reviewed (commented):
IIUC the basic fix here is track the last pc that was found before things "got stuck" so we print it. Though I'm unclear on the details - didn't we already print this pc in the stack trace as it was the last good pc? If so what does the new `print_code` show in addition?

An example of the before/after output in the hs_err file would be helpful.

Thanks.

TheRealMDoerr commented:
I noticed that we should check the result of `print_code` and update `printed_len`. Fixed with 2nd commit. Example output added to the description.

TheRealMDoerr commented:
/label add hotspot-compiler

dholmes-ora commented:
```
C  0x000001635fe021f4

called by the following code:
```
So does everything after `called by the following code` relate to the frame `C  0x000001635fe021f4`? If so I'd like to see this delineated more clearly as presently it would appear very disruptive when reading the initial sections of the hs_err file. 

Thanks

TheRealMDoerr commented:
Hi David, I've moved the verbose parts to the step "printing code blobs if possible" and updated the example output. In this particular case, the steps between the native stack and my new code don't print anything, so, the output looks very similar.

dholmes-ora commented on src/hotspot/share/utilities/vmError.cpp:
```diff
@@ -969,14 +988,15 @@ void VMError::report(outputStream* st, bool _verbose) {
-        print_code(st, _thread, lastpc, true, printed, printed_capacity);
+        const char* name = find_code_name(lastpc);
+        if (name != nullptr) {
+          st->print_cr("the last pc belongs to %s (will be printed below)", name);
```
s/the/The
s/will be//

tstuefe replied:
Why not print the nmethod right here?

dholmes-ora replied:
See my earlier comments. I was looking for some kind of delineation of all this additional stuff so that it would be easier to see where it fits into things. Not sure we've really achieved that regardless.

dholmes-ora reviewed (approved):
I guess I never come across hs_err logs with this kind of detail as I was expecting to see something between the "T H R E A D" and "P R O C E S S" sections.

A couple of minor nits but otherwise I guess this is okay.

Thanks.

tstuefe commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -1008,7 +1008,7 @@ class os: AllStatic {
 
  public:
   inline static bool platform_print_native_stack(outputStream* st, const void* context,
-                                                 char *buf, int buf_size);
```
Small nit, can we make this not a reference, but a pointer, and make it optional and possibly default it to nullptr?

TheRealMDoerr replied:
Then, we should probably name it `lastpc_ptr`? Would that really be better? I'm not conviced.

tstuefe reviewed (approved):
Looks good. Small nits only, up to you.

TheRealMDoerr commented:
Thanks for the reviews!

TheRealMDoerr commented:
/integrate
